### PR TITLE
Modular configuration management

### PR DESCRIPTION
**Pull request by dunnock**
## Description

Implement modular Config loaders using as much s possible Default, Serialize and Deserialize traits. 

Summary of restructing:
- All new changes implemented in `src/configuration/loader.rs`
- Split `configuration.rs` into `configuration/global.rs` and `configuration/utils.rs` submodules
- Moved ConfigBootsrap code from lib.rs to under `configuration/bootstrap.rs` submodule.
- It is back compatible, not removing old code, though ConfigExtractor trait marked as deprecated.

UPDATED Description 17/Apr: 
Provide simplified implementation for modular configurations, which can encapsulate structure and code when working with specific configuration. Configuration should follow same approach with overloading from network (or custom) subsections.

Tari is using config crate which allows to extend config file with application level configs. To allow deriving configuration from a Config via `ConfigLoader` trait application configuration struct should implements `Deserialize` and `NetworkConfigPath` traits.
In such case `ConfigLoader::load_from` logic will include automated overloading of parameters from [application.{network}] subsection, where network is specified in application.use_network parameter.

Trait `ConfigPath` allows to customize overloading logic even further and `DefaultConfigLoader` trait accounts for struct `Defaults` when loading values, but also requires struct to impl `Serialize`.

Details on `ConfigPath` and `NetworkConfigPath`:
```
/// Load struct from config's main section and subsecttion override
pub trait ConfigPath {
    /// Main configuration section
    fn main_key_prefix() -> &'static str;
    /// Overload values from a key prefix based on some configuration value.
    ///
    /// Should return a path to configuration table with overloading values.
    /// Returns `ConfigurationError` if key_prefix field has wrong value.
    /// Returns Ok(None) if no overload is required
    fn overload_key_prefix(config: &Config) -> Result<Option<String>, ConfigurationError>;
}

/// Load struct from config's main section and network subsection override
///
/// Network subsection will be choosen based on `use_network` key value
/// from the main section defined in this trait.
/// Wrong network value will result in Error
pub trait NetworkConfigPath {
    /// Main configuration section
    fn main_key_prefix() -> &'static str;
}
```

## Motivation and Context
As developer of custom Tari Application (Validator Node) I like to be able to define and load custom config sections easily, as well as that should follow common tari configuration approach. ConfigExtractor trait allowing to achieve that but requires quite a lot of boilerplate for deserializing data into struct as well as managing defaults.

Config crate plays nicely with Serialize and Deserialize traits, this approach is reusing serde power as much as possible to make defining and loading configuration structs as easy as possible. It also provides default parameters setting reusing Default trait.

## How Has This Been Tested?
I have covered combination of variants for config loading with tests:
- 4 traits: DefaultConfigLoader, ConfigLoader, ConfigPath, NetworkConfigPath
- Loading value from main section, network subsection, without subsection, default, default from serde
- When value / section is not defined
All tests are located in loader.rs submodule.

Manually tested node startup, with initialization from the scratch.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [x] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by CjS77** [comment]
+1 on all of this.

I was thinking on this last night and thought of the following:

We could implement a _convention_ that removes the need to specify the network (or any contingent config in general) directly.
Given a prefix key `[my_section]`, if there exists a field `mysection.use_network=NNN`, then all fields from `[my_section.NN]` get "promoted" or flattened into `[my_section]` effectively.

i.e. `my_section.NN.foo=42` will be available as `my_section.foo`.

If `mysection.use_network` does not exist, if the code supplies a default value then collapsing can still occur, otherwise no collapsing is done.

We could generalise this even further and say `my_section.use_SUBSECTION` collapses any fields from `my_section.SUBSECTION`.

**Comment by dunnock** [comment]
@CjS77  cool, I will extend PoC with TryFrom trait which will be considering your suggestions

**Comment by dunnock** [comment]
@CjS77 it's ready for 2nd review:
It is 3rd approach, it's not using ConfigExtractor trait anymore, instead provides one of 2 traits for implementation - generic one allow to implement stuff like `my_section.use_SUBSECTION` and second assumes there is `mysection.use_network` key:
```
/// Load struct from config's main section and subsecttion override
pub trait ConfigPath {
    /// Main configuration section
    fn main_key_prefix() -> &'static str;
    /// Overload values from a key prefix based on some configuration value.
    ///
    /// Should return a path to configuration table with overloading values.
    /// Returns `ConfigurationError` if key_prefix field has wrong value.
    /// Returns Ok(None) if no overload is required
    fn overload_key_prefix(config: &Config) -> Result<Option<String>, ConfigurationError>;
}

/// Load struct from config's main section and network subsection override
///
/// Network subsection will be choosen based on `use_network` key value
/// from the main section defined in this trait.
/// Wrong network value will result in Error
pub trait NetworkConfigPath {
    /// Main configuration section
    fn main_key_prefix() -> &'static str;
}
```

This grants automatically implemented 2 traits, one of those should be used to load configuration:
```
impl<C> ConfigLoader for C where C: ConfigPath + for<'de> serde::de::Deserialize<'de> {}

impl<C> DefaultConfigLoader for C where C: ConfigPath + Default + serde::ser::Serialize + for<'de> serde::de::Deserialize<'de>
```

so final code on user side is:
```
#[derive(Serialize, Deserialize)]
struct MyNodeConfig {
    welcome_message: String,
    goodbye_message: String,
}
impl Default for MyNodeConfig {
    fn default() -> Self {
        Self {
            welcome_message: "welcome from tari".into(),
            goodbye_message: "bye bye".into(),
        }
    }
}
impl NetworkConfigPath for MyNodeConfig {
    fn main_key_prefix() -> &'static str {
        "my_node"
    }
}
let mut config = Config::new();
config.set("my_node.goodbye_message", "see you later");
let my_config = MyNodeConfig::try_from(config);
assert_eq!(my_config.googbye_message, "see you later".to_string());
assert_eq!(my_config.welcome_message, "welcome from tari".to_string());
```

New traits:
https://github.com/dunnock/tari/blob/dunnock-config-extractor-auto-impl/common/src/configuration.rs#L200-L308

Example usage reimplementing mempool:
https://github.com/dunnock/tari/blob/dunnock-config-extractor-auto-impl/common/examples/mempool_config_v2.rs#L73-L130

**Comment by CjS77** [comment]
Boom! This is 🔥 🔥

**Comment by dunnock** [comment]
Cool, so I will work on proper implementation!

**Comment by dunnock** [comment]
@cjs77 should I deprecate ConfigExtractor or can replace it with new traits?

**Comment by dunnock** [comment]
Amended lots of docs, also it involved some restructuring of configuration.rs and lib.rs code - wanted to split loader.rs separately. Updated PR description, so there is most relevant info.

**Comment by CjS77** [comment]
Cool. One small bug identified, and a nit, but otherwise looks great

File: common/examples/base_node_init.rs [hunk 1]
```diff
@@ -7,13 +9,69 @@ struct Arguments {
...
+    let mut config = args.bootstrap.load_configuration()?;
+
+    // load basic config directly via Deserialize trait:
+    let basic_config: BasicConfig = config.clone().try_into()?;
+    assert_eq!(basic_config.welcome_message, welcome());
+
+    let my_config: MyNodeConfig = MyNodeConfig::load_from(&config)?;
+    assert_eq!(my_config.welcome_message, welcome());
+
+    config.set("my_node.use_netwrok", "mainnet")?;
```
[the remark this answers is not in the document]
Comment (dunnock) [reply common/examples/base_node_init.rs]: Oh indeed, sorry for that, fixed

File: common/examples/mempool_config.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    pub storage_capacity: usize,
+    /// The maximum number of transactions that can be skipped when compiling a set of highest priority transactions,
+    /// skipping over large transactions are performed in an attempt to fit more transactions into the remaining space.
+    #[serde(rename = "weight_tx_skip_count")]
+    pub weight_tx_skip_count: usize,
+}
+impl Default for UnconfirmedPoolConfig {
+    fn default() -> Self {
+        Self {
+            storage_capacity: 1024,
```
[the remark this answers is not in the document]
Comment (dunnock) [reply common/examples/mempool_config.rs]: I guess it applies only to defaults? I think fixed now

**Comment by CjS77** [comment]
Looks good; there's a linting error, which I'll fix and then merge this PR directly.

**Comment by dunnock** [comment]
Thanks and sorry for that lint, was late night fix

**Pull request merged**